tests/integration/test_task_pool.py

```diff
@@ -1224,41 +1181,391 @@ async def test_detect_incomplete_tasks(
+                'graph': {
+                    'P1': "foo\n foo[+P1] & bar => baz"
+                }
+            }
+        }
+    )
+    schd = scheduler(id_)
+    async with start(schd) as log:
+        for itask in schd.pool.get_tasks():
+            schd.pool.spawn_on_output(itask, "succeeded")
+        assert log_filter(
+            log,
+            regex=(
+                ".*1/baz.*not spawned: a prerequisite is beyond"
+                r" the workflow stop point \(1\)"
+            )
+        )
+
+
+async def test_set_failed_complete(
+    flow,
+    scheduler,
+    start,
+    one_conf,
+    log_filter,
+    db_select: Callable
+):
+    """Test manual completion of an incomplete failed task."""
+    id_ = flow(one_conf)
+    schd = scheduler(id_)
+    async with start(schd) as log:
+        one = schd.pool.get_tasks()[0]
+        one.state_reset(is_queued=False)
+
+        schd.pool.task_events_mgr.process_message(one, 1, "failed")
+        assert log_filter(
+            log, regex="1/one.* setting implied output: submitted")
+        assert log_filter(
+            log, regex="1/one.* setting implied output: started")
+        assert log_filter(
+            log, regex="failed.* did not complete required outputs")
+
+        # Set failed task complete via default "set" args.
+        schd.pool.set_prereqs_and_outputs([one.identity], None, None, ['all'])
+
+        assert log_filter(
+            log, contains=f'[{one}] task completed')
+
+        db_outputs = db_select(
+            schd, True, 'task_outputs', 'outputs',
+            **{'name': 'one'}
+        )
+        assert (
+            sorted(loads((db_outputs[0])[0])) == [
+                "failed", "started", "submitted", "succeeded"
+            ]
+        )
+
+
+async def test_set_prereqs(
+    flow,
+    scheduler,
+    start,
+    log_filter,
+):
+    """Check manual setting of prerequisites.
+
+    """
+    id_ = flow(
+        {
+            'scheduler': {
+                'allow implicit tasks': 'True',
+            },
+            'scheduling': {
+                'initial cycle point': '2040',
+                'graph': {
+                    'R1': "foo & bar & baz => qux"
+                }
+            },
+            'runtime': {
+                'foo': {
+                    'outputs': {
+                        'a': 'drugs and money',
+                    }
+                }
+            }
+        }
+    )
+    schd = scheduler(id_)
+
+    async with start(schd) as log:
+
+        # it should start up with foo, bar, baz
+        assert (
+            pool_get_task_ids(schd.pool) == ["20400101T0000Z/bar", "20400101T0000Z/baz", "20400101T0000Z/foo"]
+        )
+
+        # try to set an invalid prereq of qux
+        schd.pool.set_prereqs_and_outputs(
+            ["20400101T0000Z/qux"], None, ["20400101T0000Z/foo:a"], ['all'])
+        assert log_filter(
+            log, contains='20400101T0000Z/qux does not depend on "20400101T0000Z/foo:drugs and money"')
```

Comment:
Looks like this is logging messages rather than triggers:

```suggestion
            log, contains='20400101T0000Z/qux does not depend on "20400101T0000Z/foo:a"')
```

Reply:
Yes, last case of that. It's trickier because only task messages are available there in the code ... but I found away.

Reply:
(fixed)